lucene/core/src/java/org/apache/lucene/search/SliceExecutionControlPlane.java

```diff
@@ -0,0 +1,54 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.lucene.search;
+
+import java.util.concurrent.Executor;
+
+/**
+ * Manages the end to end execution and management of slices for an IndexSearcher
+ * instance. This includes execution on the underlying Executor and managing
+ * thread allocations to ensure a consistent throughput under varying stress loads.
+ * This class can block allocation of new slices when executing a query
+ * Remaining segments will be allocated to a single slice
+ * NOTE: This can be degrading to query performance since one thread can
+ * be overloaded with multiple segments hence this is a tradeoff
+ * between query throughput and latency
+ *
+ * A typical implementation of this interface would include the execution framework
+ * along with a circuit breaker condition which will control whether new threads will be
+ * created or not.
+ */
+public interface SliceExecutionControlPlane {
+
+  /**
+   * Return true if the circuit breaker condition has triggered,
+   * false otherwise
+   */
+  boolean hasCircuitBreakerTriggered();
```

Comment:
Maybe we should move the logic of assigning tasks to threads to this class to give it more flexibility instead of just exposing whether the pool is running over capacity. I'm thinking of something that could look like `void invokeAll(Collection<Runnable> tasks)` (similar to ForkJoinPool), which could then decide to merge some runnables together, run some of them on the current thread, etc.

Reply:
Would that also include creating the actual slices? We would need this class to be involved in that process as well since it can control the number of slices being created. Or maybe we create the slices as we do today, but can run multiple slices on caller thread when things are hot?

Reply:
I was thinking the latter indeed, having more and more slices in a single task as the number of entries in the queue increases.